src/quic/session.cc

```diff
@@ -0,0 +1,3506 @@
+  size_t len = 0;
+  for (size_t n = 0; n < count; n++) len += vec[n].len;
+  return len;
+}
+}  // namespace
+
+class DefaultApplication final : public Session::Application {
+ public:
+  DefaultApplication(Session* session,
+                     const Session::Config& config,
+                     const Application::Options& options);
+
+  QUIC_NO_COPY_OR_MOVE(DefaultApplication)
+
+  bool ReceiveStreamData(Stream* stream,
+                         ReceiveStreamDataFlags flags,
+                         const uint8_t* data,
+                         size_t datalen,
+                         uint64_t offset) override;
+
+  int GetStreamData(StreamData* stream_data) override;
+
+  void ResumeStream(stream_id id) override;
+  bool ShouldSetFin(const StreamData& stream_data) override;
+  bool StreamCommit(StreamData* stream_data, size_t datalen) override;
+
+  SET_SELF_SIZE(DefaultApplication)
+  SET_MEMORY_INFO_NAME(DefaultApplication)
+  SET_NO_MEMORY_INFO()
+
+ private:
+  void ScheduleStream(stream_id id);
+  void UnscheduleStream(stream_id id);
+
+  Stream::Queue stream_queue_;
+};
+
+DefaultApplication::DefaultApplication(Session* session,
+                                       const Session::Config& config,
+                                       const Application::Options& options)
+    : Session::Application(session, options) {}
+
+void DefaultApplication::ScheduleStream(stream_id id) {
+  auto stream = session().FindStream(id);
+  if (LIKELY(stream && !stream->is_destroyed())) {
+    DEBUG_ARGS(&session(), "Scheduling stream %" PRIi64, id);
+    stream->Schedule(&stream_queue_);
+  }
+}
+
+void DefaultApplication::UnscheduleStream(stream_id id) {
+  auto stream = session().FindStream(id);
+  if (LIKELY(stream)) {
+    DEBUG_ARGS(&session(), "Unscheduling stream %" PRIi64, id);
+    stream->Unschedule();
+  }
+}
+
+void DefaultApplication::ResumeStream(stream_id id) {
+  ScheduleStream(id);
+}
+
+bool DefaultApplication::ReceiveStreamData(Stream* stream,
+                                           ReceiveStreamDataFlags flags,
+                                           const uint8_t* data,
+                                           size_t datalen,
+                                           uint64_t offset) {
+  // One potential DOS attack vector is to send a bunch of empty stream frames
+  // to commit resources. Check that here. Essentially, we only want to create a
+  // new stream if the datalen is greater than 0, otherwise, we ignore the
+  // packet. ngtcp2 should be handling this for us, but we handle it just to be
+  // safe. We also want to make sure that the stream hasn't been destroyed.
+  if (LIKELY(datalen > 0 && !stream->is_destroyed())) {
+    DEBUG_ARGS(&session(), "Receive stream %" PRIi64 " data", stream->id());
+    stream->ReceiveData(flags, data, datalen, offset);
+  }
+  return true;
+}
+
+int DefaultApplication::GetStreamData(StreamData* stream_data) {
+  if (stream_queue_.IsEmpty()) return 0;
```

Comment:
This short circuit bypasses `stream_data->stream.reset(stream);`

My notes don't say why but this was an issue.